Forms: Add forms hidden field block - beta

enejb opened the pull request:
This PR adds a new input field block. 

That user can fill out and it gets submitted when the user fills out the form with the value that they pre-define. 

<img width="1648" height="526" alt="Screenshot 2025-08-22 at 4 54 18 PM" src="https://github.com/user-attachments/assets/4f84ab67-a54a-43ae-8e40-1f1ab255c41c" />


As a user when submitting you don't see the hidden field.
<img width="1062" height="1148" alt="Screenshot 2025-08-22 at 4 51 33 PM" src="https://github.com/user-attachments/assets/73ae2bfb-a463-4465-9b4c-5b3a84374ea3" />

When viewing the value in the backend you see it. 

<img width="1216" height="1116" alt="Screenshot 2025-08-22 at 4 51 57 PM" src="https://github.com/user-attachments/assets/1aafd296-8473-450c-b829-674785f9d86a" />



## Proposed changes:
* Add a new block that lets the user define an hidden input field. 

### Other information:

- [x] Have you written new tests for your changes, if applicable?
- [x] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [x] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
p1750782244005599-slack-C052XEUUBL4

## Does this pull request change what data or activity we track or use?
no

## Testing instructions:

* On a new page add a form. 
* Add a hidden input field block ( like any other block) 
* Does the form block show up in the form as expected? 
* Submit the form. 
* Does the field show up as expected when showing users what the user filled out. 
* Test both as a Ajax and regular field submission. 

Use the filter to change the default value. 
```
add_filter( 'jetpack_forms_enable_ajax_submission', '__return_false' );
```

enejb committed:
Forms: Add hidden input field

enejb committed:
changelog

enejb committed:
Updated design

enejb committed:
Add filter so that devs can populate the hidden field with dynamic data.

edanzer reviewed (commented):
Great to see this PR - very valuable new field addition. 

Just reviewed code and tested. Tests great with one exception: the hidden field is shown on the confirmation screen to the user after submission. That's a solid blocker to fix before merging. 

Code looks straightforward and fully inline with how we handle block registration, editor rendering, and frontend rendering.

ilonagl commented:
A few quick suggestions based on the current implementation + some inspiration from competitors:

Default naming: Maybe we default the field label to “Hidden Label,” and let users rename it (e.g., “Company ID”) to clarify what data they're storing.

Block description: Could we update the sidebar/help text to something like:
"Invisible to users, this field lets you store extra data with each form submission."
It’s a bit more user-friendly and sets the right expectation.

MVP scope: Should we consider any extra functionality for the initial release? e.g., default value or dynamic population toggle?

Competitor's implementation:
<img width="1512" alt="Screenshot 2025-06-25 at 16 55 22" src="https://github.com/user-attachments/assets/c20bf4df-e544-4a2f-9ccf-cbacc8131d17" />

CGastrell commented on projects/packages/forms/src/blocks/field-hidden/index.js:
```diff
@@ -0,0 +1,35 @@
+	},
+	example: {
+		attributes: {
+			label: __( 'ACME Inc.', 'jetpack-forms' ),
```
Might it be that these 2, label and default, are inverted on the example?

enejb replied:
Good catch!

enejb committed:
Revert the example

enejb committed:
Simply the UX again.

enejb committed:
remove the CSS

enejb committed:
Remove the Hidden Field rendering from the web and ajax contexts.

enejb committed:
Add JetpackImageSelectField back

enejb committed:
Add a test for the hidden field

enejb committed:
Add onEnter new block

enejb committed:
Add blockProps

enejb committed:
Update the description copy.

enejb committed:
Add help so that it is less confusing what the thing are once

enejb committed:
Make beta work as expected.

enejb committed:
Add Query parameter variation

simison commented on projects/packages/forms/src/blocks/field-hidden/edit.js:
```diff
@@ -0,0 +1,59 @@
+
+	let icon = unseen;
+	let label = __( 'Hidden Field', 'jetpack-forms' );
+	let valueLabel = __( 'Field Value', 'jetpack-forms' );
```
```suggestion
	let valueLabel = __( 'Field value', 'jetpack-forms' );
```

simison commented on projects/packages/forms/src/blocks/field-hidden/edit.js:
```diff
@@ -0,0 +1,59 @@
+					__next40pxDefaultSize
+					__nextHasNoMarginBottom
+					onChange={ handleLabelChange }
+					label={ __( 'Field Label', 'jetpack-forms' ) }
```
```suggestion
					label={ __( 'Field label', 'jetpack-forms' ) }
```

Let's use sentence case everywhere.

simison commented on projects/packages/forms/src/blocks/field-hidden/edit.js:
```diff
@@ -0,0 +1,59 @@
+					onChange={ handleLabelChange }
+					label={ __( 'Field Label', 'jetpack-forms' ) }
+					hideLabelFromVision={ true }
+					placeholder={ __( 'Field Label', 'jetpack-forms' ) }
```
```suggestion
					placeholder={ __( 'Field label', 'jetpack-forms' ) }
```

simison commented on projects/packages/forms/src/blocks/field-hidden/edit.js:
```diff
@@ -0,0 +1,59 @@
+					hideLabelFromVision={ true }
+					placeholder={ __( 'Field Label', 'jetpack-forms' ) }
+					value={ attributes.label }
+					help={ attributes.label ? __( 'Field Label', 'jetpack-forms' ) : '' }
```
Let's switch to normal labels: this `help` trickery is very jumpy visually and not exactly pretty with accessibility.

<img width="638" height="148" alt="image" src="https://github.com/user-attachments/assets/4db1a2cc-95f5-4bf6-b7c6-2439f3e3fe3c" />

simison commented on projects/packages/forms/src/blocks/field-hidden/index.js:
```diff
@@ -0,0 +1,41 @@
+	),
+	icon: {
+		foreground: getIconColor(),
+		src: <Icon icon={ unseen } />,
```
Not sure but I think this would work:

```suggestion
		src: unseen,
```

simison commented on projects/packages/forms/src/contact-form/class-contact-form-field.php:
```diff
@@ -1504,6 +1504,42 @@ class="jetpack-form-file-field__container"
+		 * @return string The modified value of the hidden field.
+		 */
+		$value = apply_filters( 'jetpack_forms_hidden_field_value', $value, $label, $variation, $id );
+		return "<input type='hidden' name='" . esc_attr( $id ) . "' id='" . esc_attr( $id ) . "' value='" . esc_attr( $value ) . "' $interactivity_attrs />\n";
```
What's the use of `$interactivity_attrs` in a hidden field? 🤔

enejb replied:
The $interactivity_attrs allow the values of this field to be accessible via the interactivity api store. 
This they are not being used at this point but it makes it so that in the future these values can be display or acted upon in the future or modified via a different field.

simison replied:
Wouldn't it make sense to add them in future once we use them?

Or is it kind of obscure and might be hard to find out that they need to be added?

I'd just prefer to not have code or features around "for future/in case" and always keeping it simpler when possible.

enejb replied:
I could see that we might be miss in the future and cause a bugs since it works different from all the other fields. 
but happy to remove it.

lezama committed:
Forms: Fix hidden field not appearing in responses and simplify implementation

- Move hidden field block from 'experimental' to 'beta' variation to match PR description
- Remove query parameter variation that was causing issues with responses display
- Simplify hidden field to only support static values
- Remove variations.js file and associated complexity
- Clean up edit component and remove unused imports
- Fix filter call to use 'static' instead of undefined $variation variable

The hidden field now properly appears in the admin responses inbox.

lezama committed:
Forms: Improve hidden field UI with clearer terminology and help text

- Change "Field Label" to "Field Name" for better clarity
- Add context note "This field is invisible to site visitors"
- Add descriptive help text for both fields:
  * Field Name: "Internal name used to identify this field in form responses"
  * Field Value: "The value that will be submitted with the form"
- Improve user understanding of hidden field functionality

simison committed:
HStack fields side-by-side

simison committed:
Allow hidden fields in multi-step

enejb committed:
Add a hidden field test.

enejb committed:
simply icon

enejb committed:
Revert the interactivity api and variation

simison committed:
Update block description

simison committed:
Mobile friendly

simison reviewed (dismissed):
Works well enough in my testing to merge as beta, and then after couple follow ups release to customers:

- ✅ Placeholder works on mobile and desktop
    <img width="384" height="313" alt="Screenshot 2025-09-05 at 16 04 49" src="https://github.com/user-attachments/assets/32270240-5ec2-4796-846e-880b55faa142" />
    <img width="671" height="272" alt="Screenshot 2025-09-05 at 16 04 44" src="https://github.com/user-attachments/assets/820d8af9-69d2-4322-b8c6-97bef76ef46f" />

- ✅ Submission confirmation hides value
    <img width="745" height="337" alt="Screenshot 2025-09-05 at 16 06 30" src="https://github.com/user-attachments/assets/d275c243-6aec-4037-b35d-4e0863c5a0ad" />

- ❌ Input gets saved in the output; not good, we should not store html and only <!-- -->  should be stored with attributes.
```html
<!-- wp:jetpack/field-hidden {"label":"Test","default":"123 {test} [test] \u003cu\u003euuu \u0022 ü é $ 🔥"} -->
<input type="hidden" name="Test" value="123 {test} [test] <u&gt;uuu &quot; ü é $ 🔥"/>
<!-- /wp:jetpack/field-hidden -->
```

- ✅ ❌ In the email value is visible, but I saw some bug — might've been because my value in the hidden input was trying to break things. See form above.
 
 <img width="706" height="334" alt="image" src="https://github.com/user-attachments/assets/52113876-5baf-4141-9ad7-0d5559966e76" />
The field I used:

- ✅ ❌ Value visible in inbox but same issue as with the email (follow up; add some label to differentiate from other fields?)
   <img width="704" height="467" alt="image" src="https://github.com/user-attachments/assets/5b2abf28-9375-42af-808a-473a4b9c0c32" />

- ❌ Needs follow-up but can be separate PR: when hidden field(s) are the only inputs in a form, no form renders:
    <img width="591" height="361" alt="Screenshot 2025-09-05 at 16 05 23" src="https://github.com/user-attachments/assets/702914a6-7dd5-4509-a788-1c34e94ffbff" />

- 🟨 Did not test integrations, Google sheets export and CSV export — please try those out as well!

enejb committed:
Fix missing quote.

enejb commented:
Now the hidden field does show the submit button. 

But I am a bit torn about this experience. So when the user submits a form that only has a hidden field. The form doesn't submit since we check if the form is empty. ( the user hasn't entered any data so it should be empty ) 

But I can see how this could be potentially a valid thing that the site admin wants to do. get a user to click a button. 
Have you read this terms of service. (for example)  

Adding a chekbox here would solve this issue but what if the admin doesn't want that. 

Should be add a checkbox to mark this field as non empty in the UI? cc @Automattic/zap

enejb committed:
Save empty

tbradsha commented on projects/packages/forms/src/blocks/field-hidden/index.js:
```diff
@@ -0,0 +1,37 @@
+	...defaultSettings,
+	title: __( 'Hidden field', 'jetpack-forms' ),
+	description: __(
+		'Invisible to site visitors. Allows you to store extra values with each form submission.',
```
I'd be cautious as to the wording here, as it doesn't render on the page but it's definitely possible for an end user to see it if they look at the code. That's nearly the same thing, but we'd hate for a site owner to be embarrassed because they put something they thought the end user couldn't see. 😅 

Some ways to address this:
* Mention in the panel that it's rendered as an `<input type="hidden">`
* Mention it in the documentation
* Add the field content server-side, post-submit (though I suspect the way forms work makes this not possible)
